sem/utils.py

```diff
@@ -481,3 +489,231 @@ def convert_environment_str_to_dict(log_components):
             raise ValueError("Provided log_components string '%s' is invalid\n" % log_components)
 
     return log_components_dict
+
+
+def process_logs(log_file):
+    """
+    Create a tinyDB instance, parse the log file and insert the logs to a
+    tinyDB instance (using parse_logs and insert_logs respectively).
+
+    Returns a tinydb instance containing the logs(in table 'logs') from the
+    log file.
+
+    Args:
+        log_file (string): Path to where the log file is stored
+    """
+    if not Path(log_file).exists():
+        raise FileNotFoundError("Cannot access file '%s'\n" % log_file)
+
+    logs = parse_logs(log_file)
+
+    db = TinyDB(os.path.join('/tmp/', "logs.json"),
+                storage=CachingMiddleware(JSONStorage))
+
+    insert_logs(logs, db)
+    return db
+
+
+def parse_logs(log_file):
+    """
+    Parse the logs from a log file.
+
+    Return a list of dictionary with each dictionary having the following
+    format:
+    dictionary = {
+        'Time': timestamp,
+        'Context': context/nodeId,
+        'Component': log component,
+        'Function': function name,
+        'Arguments': function arguments,
+        'Level': log level,
+        'Message': log message
+    }
+    Note: This function will skip the log lines that do not have the same
+          structure as ns-3 logs with prefix level set to prefix_all.
+
+    Args:
+        log_file (string): Path to where the log file is stored
+    """
+    log_list = []
+    with open(log_file) as f:
+        for log in f:
+            # Groups structure
+            # group[1] = Time
+            # group[2] = Context
+            # group[3] = Extended Context ; For example, '-1 [node -1]' group[2] = -1 and group [3] = [node -1]
+            # group[4] = Component:Function(Arguments)
+            # group[5] = Component
+            # group[6] = Function
+            # group[7] = Arguments
+            # group[8] = :[Level] Message
+            # group[9] = Level/'Level '
+            # group[10] = Level
+            # group[11] = Extra space after level if present;else None
+            # group[12] = Mesage
+
+            # Example: '+0.000000000s -1 PowerAdaptationDistance:SetupPhy(): [DEBUG] OfdmRate6Mbps 0.00192 6000000bps'
+            # group[1] = 0.000000000
+            # group[2] = -1
+            # group[3] = None
+            # group[4] = PowerAdaptationDistance:SetupPhy()
+            # group[5] = PowerAdaptationDistance
+            # group[6] = SetupPhy
+            # group[7] = ''
+            # group[8] = : [DEBUG] OfdmRate6Mbps 0.00192 6000000bps
+            # group[9] = DEBUG
+            # group[10] = DEBUG
+            # group[11] = None
+            # group[12] = OfdmRate6Mbps 0.00192 6000000bps
+            groups = re.match(r'\+(\d+\.\d{9})s ((?:\d+|-\d+)( \[node\ (?:\d+|-\d+)])?) (([a-zA-Z_]+):([a-zA-Z_]+)\((.*)\))(: \[((\w+)( )?)\] (.*))?', log)
```

Comment:
> Time: \+(\d+\.\d{9})s. You are assuming that it starts with a + sign (never seen a - sign, does anyone knows whether it's a thing?), at least one digit, a "." and exactly 9 digits. Please note that the time resolution can be changed in ns-3 (see here). I would propose: ([+-]?\d+\.\d+)s

I am not aware of a log starting with a '-' sign. Also, if the resolution can be modified, your suggestion seems the best option. 

> Context: ((?:\d+|-\d+)( \[node\ (?:\d+|-\d+)])?). Need explanation

(?:\d+|-\d+) - The nodeID. Can be a positive or a negative number.
( \[node\ (?:\d+|-\d+)])?) - This is to accommodate [NS_LOG_APPEND_CONTEXT](https://www.nsnam.org/doxygen/group__logging.html#gabe50035652d407c40bdaef78214c4955).

> Component: ([a-zA-Z_]+). Why not (\w+) instead? E.g., digits can be part of a component's name

No current log component of ns-3 has a digit. But technically the log component name is just a string so it makes sense to allow digits. 

> Clarification: why isn't the last group mandatory? I mean this: (: [((\w+)( )?)] (.*))?

In my simulations, it seems NS_LOG_FUNCTION does not print the '[LEVEL] Message'. This is why I made this last regex optional. 

>Also, why all the different subgroups? E.g., 4,8,9,11 from your description above

I thought it would be less error-prone to break everything into subgroups and then use the subgroups needed.

Reply:
Updated regex in [2cc7a6e](https://github.com/signetlabdei/sem/pull/51/commits/2cc7a6e9e1cc393d65621d02a9f4aabcc97425f6).